### ISSUE TITLE
Can we get a "max path weight" optional param for a_star_search?

### ISSUE THREAD

**Issue by rodya-mirov** [thebracket/bracket-lib#134]
Maybe there's another workaround.

I have a situation (I'm working through the tutorial book) where a monster can _see_ me, it's physically _near_ me, but it can't _get_ to me (typically because another monster is in the way). Like, at all; no path exists.

So it runs an A* search, and sometimes that search (depending on the size of the map and some other factors) can take seconds! It drags the game to a halt. But really, even if the thing found a (super long) path; I don't want it to take it. So what I'd like to say is, find me a path with max weight <= 20, but if not, like, go about your business. But that doesn't seem to be possible at the moment.

Based on my (partial) understanding of the A* search algorithm, this shouldn't be a significant change. Is this a reasonable thing to add? Is there another workaround that I just don't understand?

**Comment by thebracket** [comment]
That's a good idea. I'm not sure how Rust handles optional parameters (I think it doesn't, so I'll have to write a wrapper layer). You could probably fake it by counting the path steps, or making a line between the two points and making sure every cell in the line is passable (or include monsters as opaque when doing your visibility scan), but I can see that being generally useful - so I'll see what I can do.

**Comment by rodya-mirov** [comment]
Yeah, rust doesn't do optional parameters but a wrapper layer I assume would just be a few lines? I haven't looked at the original code so there could be complexity I don't see.

I'm not sure your "fake" would work in general (what if there is a path, it's just not the direct path ... bleh). And the issue is really with runtime of the A* search -- I can look at the path after it comes out and ignore it if it's too long, but if it took 3.5s to not find a path, I have a user responsivity problem 😅 

That being said I think I can replace most of my uses of A* search with Dijkstra maps, I have high hopes for how that's going to pan out.
